chain/gap/find.go

```diff
@@ -0,0 +1,147 @@
+package gap
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/sentinel-visor/chain"
+	"github.com/filecoin-project/sentinel-visor/lens"
+	"github.com/filecoin-project/sentinel-visor/model/visor"
+	"github.com/filecoin-project/sentinel-visor/storage"
+	"golang.org/x/xerrors"
+)
+
+type GapIndexer struct {
+	DB     *storage.Database
+	opener lens.APIOpener
+}
+
+func NewGapIndexer(o lens.APIOpener, db *storage.Database) *GapIndexer {
+	return &GapIndexer{
+		DB:     db,
+		opener: o,
+	}
+}
+
+func (g *GapIndexer) Run(ctx context.Context) error {
+	node, closer, err := g.opener.Open(ctx)
+	if err != nil {
+		return xerrors.Errorf("open lens: %w", err)
+	}
+	defer func() {
+		closer()
+	}()
+
+	findLog := log.With("type", "find")
+	heightGaps, err := g.detectProcessingGaps(ctx, node)
```

Comment:
I think it would be useful to identify and explicitly label null rounds. It would stop us repeatedly checking them to see if they need to be gap filled. Also it would be a useful record of the chain history. 

There are a few ways we could represent the data:
 1. use visor_processing_reports with a new status of "NULL"
 2. create a new `null_rounds` table that simply lists the heights that were determined to be null rounds
 3. create a new `height_blocks` table that contains the number of blocks at each height

I'm pretty sure that null rounds are affected by reorgs (e.g. your node happens to be following a fork with a bunch of null rounds but then discovers a better fork with a different distribution of null rounds). That implies that we need to continually re-check for null rounds until finality has been reached.... but then we should really be reprocessing tipsets after a reorg anyway.

We could perhaps handle this with a new "consensus chain" task that follows the chain 900 epochs behind the head and builds a table consisting of height, stateroot and block cids/tipset key (empty for a null round).

Reply:
Perhaps a combination of the above. I'd like to incorporate (1) into this PR since it will make gap finding more efficient.
I've filed https://github.com/filecoin-project/sentinel-visor/issues/611 for creating a canonical task for tracking null rounds.